dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/DotTemplateTool.java

```diff
@@ -89,40 +105,84 @@ public static TemplateLayout themeLayout ( String themeInode ) throws DotDataExc
-            String title = null;
-            String drawedBody;
-            if (UtilMethods.isSet(themeInode)) {
-                Identifier ident = APILocator.getIdentifierAPI().findFromInode(themeInode);
+        if(layout == null) {
+            layout = getLayout(themeInode, isPreview, getDrawedBody(themeInode, user));
+        }
+
+        return layout;
+    }
 
-                Template template = APILocator.getTemplateAPI().findWorkingTemplate(ident.getId(), sysUser, false);
+    private static DrawedBody getDrawedBody(String themeInode, User user) throws DotDataException, DotSecurityException {
+        final Identifier ident = APILocator.getIdentifierAPI().findFromInode(themeInode);
+        final Template template = APILocator.getTemplateAPI().findWorkingTemplate(ident.getId(), user, false);
 
+        return new DrawedBody(template.getTitle(), template.getDrawedBody());
+    }
+
+    private DrawedBody getDrawedBody() {
+        return new DrawedBody((String) request.getAttribute("title"), (String) request.getAttribute("designedBody"));
+    }
 
-                drawedBody = template.getDrawedBody();
-                title = template.getTitle();
+    /**
+     * Given a theme id we will parse it and return the Layout for the given template
+     *
+     * @param themeInode
+     * @param isPreview
+     * @return
+     * @throws com.dotmarketing.exception.DotDataException
+     *
+     * @throws com.dotmarketing.exception.DotSecurityException
+     *
+     */
+    public TemplateLayout themeLayout ( final String themeInode, final Boolean isPreview ) throws DotDataException, DotSecurityException {
+        final String key = themeInode + isPreview;
+        TemplateLayout layout = layoutCache.getIfPresent(key);
+
+        if(layout==null) {
+
+            final DrawedBody drawedBody;
+
+            if (UtilMethods.isSet(themeInode)) {
+                drawedBody = getDrawedBody(themeInode, sysUser);
             } else {
-                drawedBody = (String) request.getAttribute("designedBody");
-                if (request.getAttribute("title") != null) {
-                    title = (String) request.getAttribute("title");
-                }
+                drawedBody = getDrawedBody();
             }
 
-            //Parse and return the layout for this template
+            layout = getLayout(themeInode, isPreview, drawedBody);
+        }
 
-            try {
-                layout = getTemplateLayoutFromJSON(drawedBody);
-            } catch (IOException e) {
-                layout = DesignTemplateUtil.getDesignParameters(drawedBody, isPreview);
-            }
+        return layout;
+    }
 
-            layout.setTitle(title);
-            layoutCache.put(key, layout);
+    private static TemplateLayout getLayout ( final String themeInode, final Boolean isPreview,
+                                              final DrawedBody drawedBody)
+            throws DotDataException, DotSecurityException {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove the declaration of thrown exception 'com.dotmarketing.exception.DotDataException', as it cannot be thrown from method's body. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove the declaration of thrown exception 'com.dotmarketing.exception.DotSecurityException', as it cannot be thrown from method's body. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)